vm/src/stdlib/os.rs

```diff
@@ -337,6 +337,48 @@ mod _os {
         Err(vm.new_os_error("os.open not implemented on this platform".to_owned()))
     }
 
+    #[cfg(any(target_os = "linux"))]
+    #[pyfunction]
+    pub(crate) fn sendfile(
+        out_fd: i32,
+        in_fd: i32,
+        offset: i64,
+        count: u64,
+        vm: &VirtualMachine,
+    ) -> PyResult {
+        let mut file_offset = offset;
+
+        let res =
+            nix::sys::sendfile::sendfile(out_fd, in_fd, Some(&mut file_offset), count as usize)
+                .map_err(|err| err.into_pyexception(vm))?;
+        Ok(vm.ctx.new_int(res as u64))
+    }
+
+    #[cfg(any(target_os = "macos"))]
+    #[pyfunction]
+    pub(crate) fn sendfile(
+        out_fd: i32,
+        in_fd: i32,
+        offset: i64,
+        count: u64,
+        headers: OptionalArg<PyObjectRef>,
+        trailers: OptionalArg<PyObjectRef>,
+        flags: OptionalArg<i64>,
+        vm: &VirtualMachine,
+    ) -> PyResult {
+        let mut _count = count;
+        let (res, written) = nix::sys::sendfile::sendfile(
+            in_fd,
+            out_fd,
+            offset,
+            Some(&mut _count),
+            headers.into_option(),
+            trailers.into_option(),
```

Comment:
You have to process these into an array of buffers (`&[&[u8]]`) first -- something like
```rust
let headers = match headers.into_option() {
    Some(x) => Some(vm.extract_elements::<PyBytesLike>(&x)?),
    None => None,
};
let headers = headers.as_ref().map(|v| v.iter().map(|b| b.borrow_value()).collect::<Vec<_>>());
let headers = headers.as_ref().map(|v| v.iter().map(|borrowed| &**borrowed).collect::<Vec<_>>());
let headers = headers.as_deref();
```
And then the same thing for trailers. That's really inefficient, what with creating like 3 different vecs, but I think I could submit a PR to `nix` to allow sendfile to accept any `IntoIterator` over `AsRef<[u8]>`, and then we could make this more efficient later.

Reply:
So basically with those lines, we are creating bytes from the PyObject? I put those lines but now the macos tests give me this error:

```shell
error[E0277]: the trait bound `for<'r> fn(i32, i32, i64, i64, function::OptionalArg, function::OptionalArg, function::OptionalArg<i64>, &'r vm::VirtualMachine) -> std::result::Result<pyobjectrc::PyObjectRc, pyobject::PyRef<exceptions::PyBaseException>> {stdlib::os::_os::sendfile}: function::sealed::PyNativeFuncInternal<_, _, _>` is not satisfied
```

Reply:
Oh, to fix that you can add another macro call (or 2?) after line 642 in `vm/src/function.rs`

Reply:
And yep, `PyBytesLike` is anything that implements the buffer protocol (e.g. `bytes`, `bytearray`, `memoryview`) and `PyBytesLike.borrow_value()` gets a read lock to the data, and then the `&**` dereferences it to a `&[u8]`.

Reply:
Done! Now that `os` module has `sendfile` function, a python test is broken because use a `create_file` function that tries to open the file in `xb` mode, and there's no implementation for this mode.